File: languages/sample.pl
#!/usr/bin/env perl
# Perl: parse an orders CSV, total the paid ones, print a report.
use strict;
use warnings;
use List::Util qw(sum0 max);

my $REORDER_POINT = 25;
my %by_status;
my @orders;

while (my $line = <DATA>) {
    chomp $line;
    next if $line =~ /^\s*(#|$)/;
    my ($number, $total, $status) = split /,/, $line;
    push @orders, { number => $number, total => $total, status => $status };
    $by_status{$status}++;
}

my $revenue = sum0 map { $_->{total} } grep { $_->{status} eq 'paid' } @orders;
printf "%d orders, revenue %.2f, largest #%d\n", scalar @orders, $revenue,
    (sort { $b->{total} <=> $a->{total} } @orders)[0]{number};

for my $status (sort keys %by_status) {
    print "  $status: $by_status{$status}\n";
}

sub describe { my ($o) = @_; return "#$o->{number} ($o->{status})" }

__DATA__
# number,total,status
1,120.50,paid
2,42,pending
3,0,cancelled

File: languages/sample.hs
{-# LANGUAGE LambdaCase #-}
-- Haskell: a typed order pipeline with a Maybe-returning parser.
module Sample (Order (..), parseOrder, revenue, main) where

import Data.List (sortOn)
import qualified Data.Map.Strict as Map
import Text.Read (readMaybe)

data Status = Pending | Paid | Cancelled String deriving (Show, Eq)

data Order = Order
  { number :: Int
  , total  :: Double
  , status :: Status
  } deriving (Show)

parseOrder :: String -> Maybe Order
parseOrder line = case words line of
  [n, t, "paid"]    -> Order <$> readMaybe n <*> readMaybe t <*> pure Paid
  [n, t, "pending"] -> Order <$> readMaybe n <*> readMaybe t <*> pure Pending
  _                 -> Nothing

revenue :: [Order] -> Double
revenue = sum . map total . filter ((== Paid) . status)

main :: IO ()
main = do
  let orders = [o | Just o <- map parseOrder ["1 120.5 paid", "2 42 pending", "x y z"]]
      byStatus = Map.fromListWith (+) [(show (status o), 1 :: Int) | o <- orders]
  mapM_ print (sortOn number orders)
  putStrLn $ "revenue: " ++ show (revenue orders) ++ " " ++ show byStatus
